// beautiful-docs/lib/index.js
// Generated by CoffeeScript 1.4.0
(function() {
  var Generator, Manifest, async, fs, path;

  Manifest = require('./manifest');

  Generator = require('./generator');

  async = require('async');

  fs = require('fs');

  path = require('path');

  exports.version = '1.0.0';

  exports.Manifest = Manifest;

  exports.Generator = Generator;

  exports.open = function(filename, callback) {
    return Manifest.load(filename, callback);
  };

  exports.createManifestFromDir = function(dir, callback) {
    var m;
    m = new Manifest({
      title: path.basename(dir)
    });
    return fs.readdir(dir, function(err, files) {
      var f, mdFiles, _i, _len;
      mdFiles = [];
      for (_i = 0, _len = files.length; _i < _len; _i++) {
        f = files[_i];
        if (path.extname(f) === '.md') {
          mdFiles.push(path.join(dir, f));
        }
      }
      return m.addFiles(mdFiles.sort(), function(err) {
        return callback(err, m);
      });
    });
  };

  exports.generate = function(manifest, destDir, options, callback) {
    var g;
    if (options == null) {
      options = {};
    }
    if (callback == null) {
      callback = null;
    }
    g = new Generator(options);
    if (options.templatesDir) {
      return async.series([
        (function(cb) {
          return g.generate(manifest, destDir, cb);
        })
      ], function(err) {
        if (callback) {
          return callback(err);
        }
      });
    } else {
      return async.series([
        (function(cb) {
          return g.generate(manifest, destDir, cb);
        }), (function(cb) {
          return g.copyAssets(destDir, cb);
        })
      ], function(err) {
        if (callback) {
          return callback(err);
        }
      });
    }
  };

  exports.generateIndex = function(title, manifests, filename, options, callback) {
    var g;
    if (options == null) {
      options = {};
    }
    if (callback == null) {
      callback = null;
    }
    g = new Generator(options);
    return g.generateIndex(title, manifests, filename, callback);
  };

  exports.serveStaticDir = function(dir, port) {
    var app, express,
      _this = this;
    if (port == null) {
      port = 8080;
    }
    express = require('express');
    app = express();
    app.configure(function() {
      app.use(express["static"](dir));
      return app.use(express.errorHandler({
        dumpExceptions: true,
        showStack: true
      }));
    });
    return app.listen(port);
  };

}).call(this);
